manager/deallocator/deallocator.go

```diff
@@ -0,0 +1,289 @@
+package deallocator
+
+import (
+	"context"
+	"reflect"
+
+	"github.com/docker/go-events"
+	"github.com/docker/swarmkit/api"
+	"github.com/docker/swarmkit/log"
+	"github.com/docker/swarmkit/manager/state/store"
+)
+
+var (
+	resourceDeallocators = make(map[reflect.Type]resourceDeallocator)
+)
+
+func registerServiceLevelResource(resourceDeallocator resourceDeallocator) {
+	resourceDeallocators[reflect.TypeOf(resourceDeallocator.event)] = resourceDeallocator
+}
+
+// A Deallocator waits for services to fully shutdown (ie no containers left)
+// and then proceeds to deallocate service-level resources (e.g. networks),
+// and finally services themselves
+// in particular, the Deallocator should be the only place where services, or
+// service-level resources, are ever deleted!
+// it’s worth noting that this new component’s role is quite different from
+// the task reaper’s: tasks are purely internal to Swarmkit, and their status
+// is entirely managed by the system itself. In contrast, the deallocator is
+// responsible for safely deleting entities that are directly controlled by the
+// user.
```

Comment:
nit: use two newlines for paragraph breaks in comments.

Reply:
👍

Reply:
ok this is my fault for being unclear here. before there was 1 new line, and 0 blank lines. now there are 3 new lines and 2 blank lines. it just needs to be one blank line between paragraphs. sorry.

Reply:
👍